src/axom/quest/MeshTester.hpp

```diff
@@ -33,12 +61,196 @@ enum class WatertightStatus : signed char
+    const axom::IndexType nodeIdx = triCell[ n ];
+    tri[ n ][ 0 ] = x[ nodeIdx ];
+    tri[ n ][ 1 ] = y[ nodeIdx ];
+    tri[ n ][ 2 ] = z[ nodeIdx ];
+  }
+
+  return tri;
+}
 
 /// \name Mesh test and repair
 /// @{
 
 /*!
- * \brief Find self-intersections and degenerate triangles in a surface mesh.
+ * \brief Find self-intersections and degenerate triangles in a surface mesh
+ *  utilizing a Bounding Volume Hierarchy.
+ *
+ * \param [in] surface_mesh A triangle mesh in three dimensions
+ * \param [out] intersection Pairs of indices of intersecting mesh triangles
+ * \param [out] degenerateIndices indices of degenerate mesh triangles
+ *
+ * After running this function over a surface mesh, intersection will be filled
+ * with pairs of indices of intersecting triangles and degenerateIndices will
+ * be filled with the indices of the degenerate triangles in the mesh.
+ * Triangles that share vertex pairs (adjacent triangles in a watertight
+ * surface mesh) are not reported as intersecting.  Degenerate triangles
+ * are not reported as intersecting other triangles.
+ *
+ */
+#if defined(AXOM_USE_RAJA)
+template < typename ExecSpace, typename FloatType >
+void findTriMeshIntersectionsBVH(
+  mint::UnstructuredMesh< mint::SINGLE_SHAPE >* surface_mesh,
+  std::vector<std::pair<int, int> > & intersections,
+  std::vector<int> & degenerateIndices)
+{
+  SLIC_INFO("Running BVH intersection algorithm "
+            << " in execution Space: "
+            << axom::execution_space< ExecSpace >::name());
+
+  // Get allocator
+  int allocatorID = axom::execution_space< ExecSpace >::allocatorID();
+  axom::setDefaultAllocator( allocatorID );
+
+  constexpr int NDIMS = 3;
+  constexpr int stride = 2 * NDIMS;
+  const int ncells = surface_mesh->getNumberOfCells();
+
+  Triangle3* tris = axom::allocate <Triangle3> (ncells);
+
+  FloatType* xmin = axom::allocate< FloatType >( ncells );
+  FloatType* ymin = axom::allocate< FloatType >( ncells );
+  FloatType* zmin = axom::allocate< FloatType >( ncells );
+
+  FloatType* xmax = axom::allocate< FloatType >( ncells );
+  FloatType* ymax = axom::allocate< FloatType >( ncells );
+  FloatType* zmax = axom::allocate< FloatType >( ncells );
+
+  // Each access-aligned bounding box represented by 2 (x,y,z) points
+  FloatType* aabbs = axom::allocate< FloatType >( ncells * stride );
+
+  // Finding degenerate indices and bounding boxes from Triangles
+  Triangle3 t1 = Triangle3();
+  Triangle3 t2 = Triangle3();
+
+  for (int i=0 ; i < ncells ; i++)
+  {
+    t1=getMeshTriangle(i, surface_mesh);
+    tris[i] = t1;
+
+    if (t1.degenerate())
+    {
+      degenerateIndices.push_back(i);
+    }
+
+    SpatialBoundingBox triBB = compute_bounds(t1);
+
+    const IndexType offset = i * stride;
+
+    xmin[i] = aabbs[ offset ]     = triBB.getMin()[0];
+    ymin[i] = aabbs[ offset + 1 ] = triBB.getMin()[1];
+    zmin[i] = aabbs[ offset + 2 ] = triBB.getMin()[2];
+
+    xmax[i] = aabbs[ offset + 3 ] = triBB.getMax()[0];
```

Comment:
```cpp
  xmax[i] = aabbs[ offset + 3 ] = triBB.getMax()[0]; 
```
Any chance these double assignments are the cause of the AABB initialization bug?
(I'm not sure how or why that would happen though).